File: 021-merge-two-sorted-lists/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2){
    // calloc = malloc + memset
    // struct ListNode *node = (struct ListNode *)malloc(1 * sizeof(struct ListNode));
    // memset(node, 0, sizeof(struct ListNode));
    
    struct ListNode *node = (struct ListNode *)calloc(1, sizeof(struct ListNode));
    struct ListNode *head = node;
    
    while(l1 || l2) {
        struct ListNode *tmp = (struct ListNode *)calloc(1, sizeof(struct ListNode));
        if (l1 && l2) {
            if (l1->val <= l2->val) {
                tmp->val = l1->val;
                node->next = tmp;
                l1 = l1->next;
            } else {
                tmp->val = l2->val;
                node->next = tmp;
                l2 = l2->next;
            }
        } else if (l1) {
            tmp->val = l1->val;
            node->next = tmp;
            l1 = l1->next;
        } else {
            tmp->val = l2->val;
            node->next = tmp; 
            l2 = l2->next;
        }
        node = node->next;
    }
    
    return head->next;
}

File: 021-merge-two-sorted-lists/solution.php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class Solution {

    /**
     * @param ListNode $l1
     * @param ListNode $l2
     * @return ListNode
     */
    function mergeTwoLists($l1, $l2) {
                
        $node = new ListNode(0);
        $head = $node;
        while ($l1 && $l2) {
            if ($l1->val <= $l2->val) {
                $node->next = new ListNode($l1->val);
                $l1 = $l1->next;
            } else {
               $node->next = new ListNode($l2->val);
               $l2 = $l2->next;
            }
            $node = $node->next;
        }
        
        while($l1) {
            $node->next = new ListNode($l1->val);
            $node = $node->next;
            $l1 = $l1->next;
        }
        
        while($l2) {
            $node->next = new ListNode($l2->val);
            $node = $node->next;
            $l2 = $l2->next;
        }
        
        return $head->next;
    }
}
